src/main/java/us/hebi/matlab/mat/format/NumberStore.java

```diff
@@ -48,4 +52,73 @@
 
     void writeMat5(Sink sink) throws IOException;
 
+    static int hashCodeForType(@Nullable NumberStore store, boolean logical, MatlabType type) {
+        if (store == null) {
+            return 0;
+        }
+        int hash = 1;
+        if (logical) {
+            for (int i = 0; i < store.getNumElements(); ++i) {
+                hash = 31 * hash + (Casts.logical(store.getDouble(i)) ? 1 : 0);
+            }
+        } else {
+            if (type == MatlabType.Single || type == MatlabType.Double) {
+                for (int i = 0; i < store.getNumElements(); ++i) {
+                    hash = 31 * hash + Double.hashCode(store.getDouble(i));
+                }
+            } else {
+                for (int i = 0; i < store.getNumElements(); ++i) {
+                    hash = 31 * hash + Long.hashCode(store.getLong(i));
+                }
+            }
+        }
+        return hash;
+    }
+
+    static boolean equalForType(@Nullable NumberStore a, @Nullable NumberStore b, boolean logical, MatlabType type) {
+        if ((a == null) != (b == null)) {
+            //  null mismatch is easy
+            return false;
+        } else if (a == null) {
+            // they're both null
+            return true;
+        }
+        if (a instanceof UniversalNumberStore && b instanceof UniversalNumberStore) {
+            UniversalNumberStore aCast = (UniversalNumberStore) a;
+            UniversalNumberStore bCast = (UniversalNumberStore) b;
+            if (aCast.type == bCast.type) {
+                // when their types are equal, then their binary content must be exactly the same (much faster)
+                return aCast.buffer.equals(bCast.buffer);
+            }
+        }
+        if (logical) {
```

Comment:
May result in different behavior for non-`UniversalNumberStore`, i.e., logical types may result in equal results even if the bytes in the buffer are different.

Reply:
This is intentional.  I have two MAT-files that load identically into MATLAB, but have different bytes.  They are both `double` in MATLAB, but one is backed by `uint8`, and the other by `double` in the data of the MAT-File.  Don't know how they got generated this way, but they're there :)  I'd love to contibute them as testcases, but they have customer data, and I don't know how to make them have different backing datatypes.

I considered removing the "fastcase" (premature optimization?) but it's so easy I left it in.

Reply:
I understood the intention, but I apparently missed the `if (aCast.type == bCast.type)` line. Your implementation is correct and makes sense.